src/Algebra/Reasoning/Semigroup.agda

```diff
@@ -0,0 +1,157 @@
+  the way to the left, to 'ε' associated all the way to the right. Then,
+  'assoc²XY' is the proof that X is equal to Y. Explicitly:
+
+  α = ((a ∙ b) ∙ c) ∙ d
+  β = (a ∙ (b ∙ c)) ∙ d
+  γ = (a ∙ b) ∙ (c ∙ d)
+  δ = a ∙ ((b ∙ c) ∙ d)
+  ε = a ∙ (b ∙ (c ∙ d))
+
+  Only reassociations that need two assoc steps are defined here.
+  -}
+  assoc²αδ : ((a ∙ b) ∙ c) ∙ d ≈ a ∙ ((b ∙ c) ∙ d)
+  assoc²αδ = trans (∙-cong (assoc a b c) refl) (assoc a (b ∙ c) d)
+
+  assoc²αε : ((a ∙ b) ∙ c) ∙ d ≈ a ∙ (b ∙ (c ∙ d))
+  assoc²αε = trans (assoc (a ∙ b) c d) (assoc a b (c ∙ d))
+
+  assoc²βγ : (a ∙ (b ∙ c)) ∙ d ≈ (a ∙ b) ∙ (c ∙ d)
+  assoc²βγ = trans (sym (∙-cong (assoc a b c) refl)) (assoc (a ∙ b) c d)
+
+  assoc²βε : (a ∙ (b ∙ c)) ∙ d ≈ a ∙ (b ∙ (c ∙ d))
+  assoc²βε = trans (assoc a (b ∙ c) d) (∙-cong refl (assoc b c d))
+
+  assoc²γβ : (a ∙ b) ∙ (c ∙ d) ≈ (a ∙ (b ∙ c)) ∙ d
+  assoc²γβ = trans (sym (assoc (a ∙ b) c d)) (∙-cong (assoc a b c) refl)
+
+  assoc²γδ : (a ∙ b) ∙ (c ∙ d) ≈ a ∙ ((b ∙ c) ∙ d)
+  assoc²γδ = begin
+    (a ∙ b) ∙ (c ∙ d) ≈⟨ assoc a b (c ∙ d) ⟩
+    a ∙ (b ∙ (c ∙ d)) ≈⟨ ∙-cong refl (sym (assoc b c d)) ⟩
+    a ∙ ((b ∙ c) ∙ d) ∎
+
+  assoc²δα : a ∙ ((b ∙ c) ∙ d) ≈ ((a ∙ b) ∙ c) ∙ d
+  assoc²δα = sym (trans (∙-cong (assoc a b c) refl) (assoc a (b ∙ c) d))
+
+  assoc²δγ : a ∙ ((b ∙ c) ∙ d) ≈ (a ∙ b) ∙ (c ∙ d)
+  assoc²δγ = begin
+    a ∙ ((b ∙ c) ∙ d) ≈⟨ ∙-cong refl (assoc b c d) ⟩
+    a ∙ (b ∙ (c ∙ d)) ≈⟨ sym (assoc a b (c ∙ d)) ⟩
+    (a ∙ b) ∙ (c ∙ d) ∎
+
+  assoc²εα : a ∙ (b ∙ (c ∙ d)) ≈ ((a ∙ b) ∙ c) ∙ d
+  assoc²εα = sym (trans (assoc (a ∙ b) c d) (assoc a b (c ∙ d)))
+
+  assoc²εβ : a ∙ (b ∙ (c ∙ d)) ≈ (a ∙ (b ∙ c)) ∙ d
+  assoc²εβ = sym (trans (assoc a (b ∙ c) d) (∙-cong refl (assoc b c d)))
+
+open Assoc4 public
+
+module Pulls (ab≡c : a ∙ b ≈ c) where
+    pullʳ : ∀ {x} → (x ∙ a) ∙ b ≈ x ∙ c
+    pullʳ {x = x} = begin
+        (x ∙ a) ∙ b ≈⟨ assoc x a b ⟩
+        x ∙ (a ∙ b) ≈⟨ ∙-cong refl ab≡c ⟩
+        x ∙ c       ∎
+
+    pullˡ : ∀ {x} → a ∙ (b ∙ x) ≈ c ∙ x
+    pullˡ {x = x} = begin
+        a ∙ (b ∙ x) ≈⟨ sym (assoc a b x) ⟩
+        (a ∙ b) ∙ x ≈⟨ ∙-cong ab≡c refl ⟩
+        c ∙ x       ∎
+
+    pull-first : ∀ {x y} → a ∙ ((b ∙ x) ∙ y) ≈ c ∙ (x ∙ y)
+    pull-first {x = x} {y = y} = begin
+      a ∙ ((b ∙ x) ∙ y) ≈⟨ ∙-cong refl (assoc b x y) ⟩
+      a ∙ (b ∙ (x ∙ y)) ≈⟨ pullˡ ⟩
+      c ∙ (x ∙ y)       ∎
+
+    pull-center : ∀ {x y} → x ∙ (a ∙ (b ∙ y)) ≈ x ∙ (c ∙ y)
+    pull-center {x = x} {y = y} = ∙-cong refl (pullˡ)
+
+    -- could be called pull₃ʳ
+    pull-last : ∀ {x y} → (x ∙ (y ∙ a)) ∙ b ≈ x ∙ (y ∙ c)
+    pull-last {x = x} {y = y} = begin
+      (x ∙ (y ∙ a)) ∙ b ≈⟨ assoc x (y ∙ a) b ⟩
+      x ∙ ((y ∙ a) ∙ b) ≈⟨ ∙-cong refl (pullʳ {x = y}) ⟩
+      x ∙ (y ∙ c)       ∎
+
+open Pulls public
+
+module Pushes (ab≡c : a ∙ b ≈ c) where
+    pushʳ : x ∙ c ≈ (x ∙ a) ∙ b
+    pushʳ {x = x} = begin
+        x ∙ c       ≈⟨ sym (∙-cong refl ab≡c) ⟩
+        x ∙ (a ∙ b) ≈⟨ sym (assoc x a b) ⟩
+        (x ∙ a) ∙ b ∎
```

Comment:
Successive appeals to `sym` suggest that it should be lifted out (permutation of `sym` through proofs), but in fact doing so quickly reveals the following:
```suggestion
 pushʳ = sym (pullʳ ab≡c)
```